AVRO-1751: Add python3 compatibility

gabriel-tincu opened the pull request:
Keeping in line with the same logic present in the avro java library
that handles type promotions, this aims to be a faithful adaptation of
that code in python.

Most tests were also ported, with the exception of the ones dealing with
enum default, since the python api does not support that yet

See :
- https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/SchemaCompatibility.java
- https://issues.apache.org/jira/browse/AVRO-1751

kojiromike commented:
Please be aware the py3 library is deprecated. The py lib has full support for modern python 3, and tests in multiple versions of python. Would it be possible to do this change in that library instead?

gabriel-tincu commented:
> Please be aware the py3 library is deprecated. The py lib has full support for modern python 3, and tests in multiple versions of python. Would it be possible to do this change in that library instead?

done

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+    missing_union_branch = "missing_union_branch"
+
+
+class AvroRuntimeException(Exception):
```
Exceptions should be declared in lang/py/avro/errors.py so that they can be identified and handled without import loops. Exceptions should also ideally inherit from more base Avro exception types.

gabriel-tincu replied:
moved

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+        locations: Optional[Set[str]] = None,
+    ):
+        self.locations = locations if locations else {"/"}
+        self.messages = messages if messages else set()
```
Curious why you use the ternary for the first two here, but `or X` for the last one.

gabriel-tincu replied:
fixed inconsistent use

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+
+    def merged_with(self, that):
+        that = cast(SchemaCompatibilityResult, that)
+        merged = copy(self.incompatibilities)
```
Consider `merged = [ *self.incompatibilities, *that.incompatibilities ]`

gabriel-tincu replied:
updated

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+        )
+
+    @staticmethod
+    def compatible():
```
I think we should avoid static methods. Instead, consider defining a module constant for COMPATIBLE that any other module can import.

gabriel-tincu replied:
Removed usage of both static methods

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+        return SchemaCompatibilityResult(SchemaCompatibilityType.compatible)
+
+    @staticmethod
+    def incompatible(incompat_type: SchemaIncompatibilityType, message: str, location: List[str]):
```
I'd prefer to avoid static methods. Could this be a module-level function instead?

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+                SchemaType.NULL, SchemaType.BOOLEAN, SchemaType.INT,
+                SchemaType.LONG, SchemaType.FLOAT, SchemaType.DOUBLE,
+                SchemaType.BYTES, SchemaType.STRING
+            }:
```
Should probably just have a module constant for the set of primitive types.

gabriel-tincu replied:
Added as module constant

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,318 @@
+            self.memoize_map[pair] = SchemaCompatibilityResult()
+            result = self.calculate_compatibility(reader, writer, location)
+            self.memoize_map[pair] = result
+        location.pop()
```
Is there any way to make mutating the location list more visible? I have the sense that this hidden mutation can lead to debuggability problems.

gabriel-tincu replied:
Other than using a context manager (which seems to be overkill) i really can't think of any way other than the present (using a collection)

kojiromike replied:
It's just my opinion, but I think when objects are mutated it's easier to miss things and make mistakes, like

```
somelist = ['a', 'b', 'c']
procedure(somelist)
value = somelist[0]  # May raise an IndexError!
```

Of course you can look in procedure to see if it's changing `list`, but if it's written like

```
somelist = ['a', 'b', 'c']
newlist = somefunc(somelist)
value = newlist[0]
```

It's a more explicit signal about what's happening to the data.

gabriel-tincu replied:
I opted to pass a modified list downstream. That way the code size kind of stays the same and it's still readable IMO

kojiromike reviewed (commented):
This is great, thanks for taking it on! I have a few thoughts I think can improve it:

1. Even though this is a port of a Java implementation, I don't think we need the static methods -- those aren't necessary in Python. Instead, consider making them "just functions" in the module.
1. The major parts of the implementation would be much better if they had some docstrings explaining them.
1. Mutating 'location' in a few places is a little obscure. Ideally I'd prefer to avoid mutation altogether, but it has its uses. If it's unavoidable, then if there's a way to make it visible and easy to trace it will really help debugging later on.

gabriel-tincu commented:
> This is great, thanks for taking it on! I have a few thoughts I think can improve it:
> 
>     1. Even though this is a port of a Java implementation, I don't think we need the static methods -- those aren't necessary in Python. Instead, consider making them "just functions" in the module.
> 
>     2. The major parts of the implementation would be much better if they had some docstrings explaining them.
> 
>     3. Mutating 'location' in a few places is a little obscure. Ideally I'd prefer to avoid mutation altogether, but it has its uses. If it's unavoidable, then if there's a way to make it visible and easy to trace it will really help debugging later on.

Dropped the static methods, addressed most of the PR comments. Added some docstrings lifted from the java codebase (on the more meatier methods). Also added some comments wrt the location issue

kojiromike commented on lang/py/avro/compatibility.py:
```diff
@@ -0,0 +1,314 @@
+
+    # pylSchemaType.INT: enable=too-many-return-statements
+
+    @staticmethod
```
There still seem to be some staticmethods in here :/

gabriel-tincu replied:
sorry... missed those ones...  fixed now

kojiromike commented on lang/py/avro/errors.py:
```diff
@@ -83,3 +83,7 @@ class UnsupportedCodec(NotImplementedError, AvroException):
     """An exception raised when incorrect arguments were passed."""
+
+
+class AvroRuntimeException(AvroException):
```
Can this be a subclass of both `RuntimeError` and `AvroException` like `UsageError` above?

gabriel-tincu replied:
fixed

kojiromike commented on lang/py/avro/test/test_compatibility.py:
```diff
@@ -0,0 +1,665 @@
+UNION_INT_ARRAY_INT = UnionSchema([{"type": SchemaType.INT}, INT_ARRAY_SCHEMA.to_json()])
+UNION_INT_MAP_INT = UnionSchema([{"type": SchemaType.INT}, INT_MAP_SCHEMA.to_json()])
+UNION_INT_NULL = UnionSchema([{"type": SchemaType.INT}, {"type": SchemaType.NULL}])
+FIXED_4_ANOTHER_NAME = parse(json.dumps({"type": SchemaType.FIXED, "name": "AnotherName", "size": 4}))
```
These are beautifully easy-to-read tests!

kojiromike commented on lang/py/avro/errors.py:
```diff
@@ -83,3 +83,7 @@ class UnsupportedCodec(NotImplementedError, AvroException):
+
+
+class AvroRuntimeException(RuntimeError, AvroException):
+    """Raised then compatibility parsing encounters an unknown type"""
```
Should this say "when"?

gabriel-tincu replied:
Fixed

kojiromike reviewed (approved):
There's a minor typo in a docstring; otherwise I think this is looking good.

kojiromike commented:
I'm good with the code. Rat, however, is failing the tests because the code doesn't have the approved license headers. Please include the header from the top of errors.py at the top of the other files.

gabriel-tincu commented:
> I'm good with the code. [Rat](https://creadur.apache.org/rat/), however, is failing the tests because the code doesn't have the approved license headers. Please include the header from the top of errors.py at the top of the other files.

@kojiromike I updated the 2 new files with the same header as `errors.py` but the pipeline is still red. After some scrolling around i bumped into this
```
[ERROR] Failed to execute goal on project trevni-avro: Could not resolve dependencies for project org.apache.avro:trevni-avro:jar:1.11.0-SNAPSHOT: Could not transfer artifact org.apache.commons:commons-lang3:jar:3.7 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]
```

Is this a transient error? Would rerunning / repushing have a chance of getting past ?

kojiromike commented:
@gabriel-tincu repushing might help, but I wouldn't keep trying if it doesn't work. Maybe @Fokko or @RyanSkraba have some idea? It doesn't seem to have anything to do with this PR...

RyanSkraba commented:
Hello!  I retriggered travis and I'll watch this -- it looks like travis occasionally experiences broken connections to maven central since the switch to travis-ci.com!  We used to see this frequently, but not ALL the time.  I'm not sure what we can do in the short term...

gabriel-tincu commented:
> Hello! I retriggered travis and I'll watch this -- it looks like travis occasionally experiences broken connections to maven central since the switch to travis-ci.com! We used to see this frequently, but not ALL the time. I'm not sure what we can do in the short term...

I guess i can also make some mock pushes along the way. Have you guys considered using github workflows? Also, are broken connections due to rate limiting on the maven side or something else? Thanks for looking into this !

RyanSkraba commented:
Hello!  For the Github Actions issue -- I raised https://issues.apache.org/jira/browse/AVRO-2991 to discuss the impact, advantages and disadvantages!   For the unreliable Travis <-> maven repo connections, I raised https://issues.apache.org/jira/browse/AVRO-2992.  We might be able to fix this quickly by setting up a maven cache.

In the meantime, I'll hit the rebuild button until I see some green (or at least some non-Java/Maven errors on your python PR!)  If you're interested, it's really quite easy to set up travis-ci.org on your own fork (so you have your own Rebuild button).

This is really unfortunate!  Thanks for your patience :/

RyanSkraba commented:
Hello!  I've definitely confirmed that the build on master is failing _in addition to_ the maven central connectivity issues!  I'm looking at this and raised one more issue: https://issues.apache.org/jira/browse/AVRO-2994

You will not get a green build on Travis until that's fixed :/

In the meantime, I manually patched the docker image locally and tried a local build :x: .  It looks like it's failing in python 3.5 because of the f-string syntax !

```
======================================================================
ERROR: avro.test.test_compatibility (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: avro.test.test_compatibility
Traceback (most recent call last):
  File "/usr/lib/python3.5/unittest/loader.py", line 428, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.5/unittest/loader.py", line 369, in _get_module_from_name
    __import__(name)
  File "/home/rskraba/avro/lang/py/avro/test/test_compatibility.py", line 565
    assert location in result.locations, f"expected {location}, found {result.location}"
                                                                                       ^
SyntaxError: invalid syntax
```

Python 3.5 reached end-of-life fairly recently if I remember correctly, but we're still supporting it.

gabriel-tincu commented:
> Hello! I've definitely confirmed that the build on master is failing _in addition to_ the maven central connectivity issues! I'm looking at this and raised one more issue: https://issues.apache.org/jira/browse/AVRO-2994
> 
> You will not get a green build on Travis until that's fixed :/
> 
> In the meantime, I manually patched the docker image locally and tried a local build x . It looks like it's failing in python 3.5 because of the f-string syntax !
> 
> ```
> ======================================================================
> ERROR: avro.test.test_compatibility (unittest.loader._FailedTest)
> ----------------------------------------------------------------------
> ImportError: Failed to import test module: avro.test.test_compatibility
> Traceback (most recent call last):
>   File "/usr/lib/python3.5/unittest/loader.py", line 428, in _find_test_path
>     module = self._get_module_from_name(name)
>   File "/usr/lib/python3.5/unittest/loader.py", line 369, in _get_module_from_name
>     __import__(name)
>   File "/home/rskraba/avro/lang/py/avro/test/test_compatibility.py", line 565
>     assert location in result.locations, f"expected {location}, found {result.location}"
>                                                                                        ^
> SyntaxError: invalid syntax
> ```
> 
> Python 3.5 reached end-of-life fairly recently if I remember correctly, but we're still supporting it.

@RyanSkraba Sorry about that, I missed that particular reference to a stale class member. For some reason (probably because the test passes) that did not blow up locally

kojiromike commented:
@RyanSkraba by the way, pip compatibility with python 3.5 ends in January. So… do we need to hold another vote, or do we have a simple process for ending support for dead snakes?

RyanSkraba commented:
That's an excellent point!  Does that mean that pip will just stop working with python 3.5 installations in January?  I know that the Databricks LTS expects python 3.5 to work until mid-2021!  (That being said, it also comes with a version of Avro pre-installed, so it's probably not our problem).

My opinion is that when pip drops support for an end-of-life release, we should also drop support!  Maybe we should just create a JIRA every time, and consider the PR that updates the build environment to be the "community vote"?

kojiromike commented:
> Maybe we should just create a JIRA every time, and consider the PR that updates the build environment to be the "community vote"?

That makes sense to me.

gabriel-tincu commented:
@RyanSkraba Made it work for 3.5 ... give it another look please

ivanyu commented:
Hi @kojiromike 
Would it be possible to merge this?
